src/scheduler/MpiWorld.cpp

```diff
@@ -1732,4 +1759,85 @@ void MpiWorld::checkRanksRange(int sendRank, int recvRank)
         throw std::runtime_error("Recv rank outside range");
     }
 }
+
+// This method will either (i) do nothing or (ii) run the whole migration.
+// Note that every rank will call this method.
+void MpiWorld::tryMigrate(int thisRank)
+{
+    auto pendingMigrations =
+      getScheduler().canAppBeMigrated(thisRankMsg->appid());
+
+    if (pendingMigrations == nullptr) {
+        return;
+    }
+
+    prepareMigration(thisRank, pendingMigrations);
+    if (faabric::util::isMockMode()) {
+        // When mocking in the tests, the call to get the current executing
+        // executor may fail
+        faabric::util::UniqueLock lock(mockMutex);
+        mpiMockedPendingMigrations.push_back(pendingMigrations);
+    } else {
+        // An executing thread may never return from the next call. However,
```

Comment:
> An executing thread may never return from the next call.

This seems like very strange behaviour and I'm not sure how this would be implemented. I would instead have `tryMigrate` return a `bool`, then if it's `true` it means it's been migrated and the caller (e.g. the MPI function) has to deal with it and ensure a graceful shutdown. 

This would probably best be done by throwing an `ExecutorMigrated` exception, which the main `Executor` loop could catch and self-terminate accordingly (as it does already with exceptions [here](https://github.com/faasm/faabric/blob/master/src/scheduler/Executor.cpp#L531)).